src/v/rpc/connection_cache.cc

```diff
@@ -100,20 +100,53 @@ ss::future<> connection_cache::stop() {
 }
 
 ss::shard_id connection_cache::shard_for(
-  model::node_id self,
+  model::node_id src_node,
   ss::shard_id src_shard,
-  model::node_id n,
+  model::node_id dest_node,
   ss::shard_id total_shards) const {
     if (total_shards <= _max_connections) {
         return src_shard;
     }
 
     // NOLINTNEXTLINE
-    size_t h = 805306457;
-    boost::hash_combine(h, jump_consistent_hash(src_shard, _max_connections));
-    boost::hash_combine(h, std::hash<model::node_id>{}(n));
-    boost::hash_combine(h, std::hash<model::node_id>{}(self));
-    // use self node id to shift jump_consistent_hash_assignment
-    return jump_consistent_hash(h, total_shards);
+    uint64_t connection_hash = 201326611; // magic prime
+    boost::hash_combine(connection_hash, std::hash<model::node_id>{}(src_node));
+    boost::hash_combine(
+      connection_hash, std::hash<model::node_id>{}(dest_node));
+
+    // Find the largest ring_size s.t. ring_size <= total_shards
+    // and ring_size % connections_per_node == 0
+    uint64_t remainder = total_shards % _max_connections;
+    uint64_t ring_size = total_shards - remainder;
+
+    // Vary the starting shard of the ring depending on the connection hash.
+    uint64_t ring_start_shard = jump_consistent_hash(
+      connection_hash, remainder + 1);
+    uint64_t ring_end_shard = (ring_size - 1) + ring_start_shard;
+
+    // Hash any shards not in [ring_start_shard, ring_end_shard] to a point
```

Comment:
So this happens because of the truncation above right? For example, 11 shards and 8 connections then we handle shard 9,10,11 here?

Reply:
Yep exactly, in any case where shard count isn't a multiple of the number of connections this will be used.